components/fxa-client/src/jwcrypto.rs

```diff
@@ -0,0 +1,261 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// Theorically, everything done in this file could and should be done in a JWT library.
+/// However, none of the existing rust JWT libraries can handle ECDH-ES encryption, and API choices
+/// made by their authors make it difficult to add this feature.
+/// In the past, we chose cjose to do that job, but it added three C dependencies to build and link
+/// against: jansson, openssl and cjose itself.
+use crate::error::*;
+use rc_crypto::{
+    aead, agreement,
+    agreement::{Ephemeral, InputKeyMaterial, KeyPair},
+    digest, rand,
+};
+use serde_derive::{Deserialize, Serialize};
+
+pub enum EncryptionParameters {
+    // ECDH-ES in Direct Key Agreement mode.
+    #[allow(non_camel_case_types)]
+    ECDH_ES {
+        enc: EncryptionAlgorithm,
+        peer_jwk: Jwk,
+        local_key_pair: KeyPair<Ephemeral>,
+    },
+}
+
+pub enum DecryptionParameters {
+    // ECDH-ES in Direct Key Agreement mode.
+    #[allow(non_camel_case_types)]
+    ECDH_ES { local_key_pair: KeyPair<Ephemeral> },
+}
+
+#[derive(Serialize, Deserialize, PartialEq)]
+enum Algorithm {
+    #[serde(rename = "ECDH-ES")]
+    #[allow(non_camel_case_types)]
+    ECDH_ES,
+}
+#[derive(Serialize, Deserialize, PartialEq)]
+pub enum EncryptionAlgorithm {
+    A256GCM,
+}
+
+impl EncryptionAlgorithm {
+    fn algorithm_id(&self) -> &'static str {
+        match self {
+            Self::A256GCM => "A256GCM",
+        }
+    }
+}
```

Comment:
This is the start of a more generic JWE API, even thought we only support our use case (ECDH-ES w/ direct key agreement and A256GCM content). It kinda sucks and is probably over-engineered, but it's a start.